### DISCUSSION TITLE
CKAN Integration Test/Benchmark Suite

### DISCUSSION THREAD

**Discussion by jqnatividad** [ckan/ckan#6424] [Ideas]
It would be nice if CKAN had a standard integration test/benchmark suite.  With it, we can:

- create a performance baseline
- have a "de facto" systems integration test 
as we discuss unit test approach (#6373), a benchmark suite can also serve as an integration test, and could be made part of the release checklist.
- support performance regression testing
with the recent Python 3/Flask migration effort largely done for CKAN core, the pace of contributions has increased and it would be nice to have an objective way to measure the impact of architectural choices as we continue to modernize the platform.

Also, because of the highly extensible nature of CKAN, it will be a good tool to objectively measure and isolate undesired side effects of extensions.

Various test types come to mind:
- Front-end testing
Using the standard templates, test with a tool like Selenium
- API testing
Using a tool like Postman not only helps with API Testing, it can make the API more accessible to downstream API users and extension developers as Postman is also widely used by developers.
- Database testing
   - CKAN system database (PostgreSQL)
   - Datastore (not only the existing PostgreSQL backend, but alternate backends as well)
- Harvester testing
  - the harvester test suite can also help with the tests above.  We can use it to distribute standard test/benchmark data at a central CKAN test instance.
  - we can even use it to help distribute more "realistic" [seed data](https://docs.ckan.org/en/2.9/maintaining/getting-started.html#creating-test-data) for testing new installations.

**Comment by Zharktas** [comment]
We already have [cypress](https://www.cypress.io/) as testing tool for JavaScript code, I've been planning to extend its use to actual front-end testing, just need to design how to populate data in consistent and fast manner for tests.
